File: scripts/lib/github-client.js
// ============================================================
// SHARED GITHUB CLIENT — Reusable helper used by ALL agents.
//
// WHY THIS EXISTS:
// Every agent needs to talk to the GitHub API (read issues,
// create branches, open PRs, post comments). Instead of
// duplicating that setup code in every script, we centralize
// it here. This follows the DRY principle.
//
// WHAT IT PROVIDES:
// - Pre-configured Octokit instance (GitHub SDK)
// - Helper functions for common operations:
//     • getRepoInfo()       — parse owner/repo from env
//     • createBranch()      — create a new git branch
//     • createOrUpdateFile() — commit a file to a branch
//     • createPullRequest() — open a PR
//     • addComment()        — post a comment on an issue/PR
//     • addLabels()         — add labels to an issue/PR
// ============================================================

import { Octokit } from "@octokit/rest";

// -----------------------------------------------------------
// Initialize Octokit with the GitHub token from environment.
//
// GITHUB_TOKEN is provided by:
// - GitHub Actions (automatically via `secrets.GITHUB_TOKEN`)
// - Your local .env file (for testing locally)
// -----------------------------------------------------------
const octokit = new Octokit({
  auth: process.env.GITHUB_TOKEN,
});

// -----------------------------------------------------------
// getRepoInfo() — Extracts owner and repo name from the
// GITHUB_REPOSITORY environment variable.
//
// GitHub Actions automatically sets this to "owner/repo".
// For local development, you set it in your .env file.
//
// Example: "octocat/hello-world" → { owner: "octocat", repo: "hello-world" }
// -----------------------------------------------------------
export function getRepoInfo() {
  const [owner, repo] = process.env.GITHUB_REPOSITORY.split("/");
  if (!owner || !repo) {
    throw new Error(
      'GITHUB_REPOSITORY must be in "owner/repo" format. ' +
        `Got: "${process.env.GITHUB_REPOSITORY}"`
    );
  }
  return { owner, repo };
}

// -----------------------------------------------------------
// getDefaultBranch() — Finds the repo's default branch name.
//
// Most repos use "main", but some older ones use "master".
// Instead of hardcoding, we ask the GitHub API.
// -----------------------------------------------------------
export async function getDefaultBranch() {
  const { owner, repo } = getRepoInfo();
  const { data } = await octokit.rest.repos.get({ owner, repo });
  return data.default_branch;
}

// -----------------------------------------------------------
// createBranch(branchName, fromBranch) — Creates a new branch.
//
// HOW IT WORKS:
// 1. Gets the latest commit SHA from the source branch
// 2. Creates a new git reference pointing to that SHA
//
// This is equivalent to: git checkout -b branchName fromBranch
//
// PARAMETERS:
//   branchName — name of the new branch (e.g., "fix/issue-42")
//   fromBranch — base branch to branch from (default: repo's default)
// -----------------------------------------------------------
export async function createBranch(branchName, fromBranch) {
  const { owner, repo } = getRepoInfo();

  // If no base branch specified, use the repo's default
  const baseBranch = fromBranch || (await getDefaultBranch());

  // Step 1: Get the SHA (commit hash) of the latest commit on the base branch
  const { data: refData } = await octokit.rest.git.getRef({
    owner,
    repo,
    ref: `heads/${baseBranch}`,
  });

  // Step 2: Create a new branch reference pointing to that same commit
  // The new branch starts as an exact copy of the base branch.
  await octokit.rest.git.createRef({
    owner,
    repo,
    ref: `refs/heads/${branchName}`,
    sha: refData.object.sha,
  });

  console.log(`✅ Branch created: ${branchName} (from ${baseBranch})`);
  return branchName;
}

// -----------------------------------------------------------
// createOrUpdateFile(branch, path, content, message)
//
// Commits a single file to a branch via the GitHub API.
// If the file already exists, it updates it; otherwise creates it.
//
// HOW IT WORKS:
// - GitHub's Contents API lets you create/update files with a
//   single API call — no need to clone the repo locally.
// - If updating, you must provide the file's current SHA so
//   GitHub knows which version you're replacing (prevents conflicts).
//
// PARAMETERS:
//   branch  — target branch to commit to
//   path    — file path in the repo (e.g., "src/utils.js")
//   content — the full file content (will be base64 encoded)
//   message — git commit message
// -----------------------------------------------------------
export async function createOrUpdateFile(branch, path, content, message) {
  const { owner, repo } = getRepoInfo();

  // Check if the file already exists (we need its SHA to update it)
  let existingSha;
  try {
    const { data } = await octokit.rest.repos.getContent({
      owner,
      repo,
      path,
      ref: branch,
    });
    existingSha = data.sha;
  } catch (error) {
    // File doesn't exist yet — that's fine, we'll create it
    if (error.status !== 404) throw error;
  }

  // Create or update the file
  // The GitHub API requires content to be Base64 encoded
  await octokit.rest.repos.createOrUpdateFileContents({
    owner,
    repo,
    path,
    message,
    content: Buffer.from(content).toString("base64"),
    branch,
    ...(existingSha && { sha: existingSha }),
  });

  console.log(`✅ File committed: ${path} → ${branch}`);
}

// -----------------------------------------------------------
// getFileContent(path, branch) — Read a file's content from the repo.
//
// Returns the decoded (UTF-8) content of the file.
// Useful for agents that need to read existing code before modifying it.
// -----------------------------------------------------------
export async function getFileContent(path, branch) {
  const { owner, repo } = getRepoInfo();

  const { data } = await octokit.rest.repos.getContent({
    owner,
    repo,
    path,
    ref: branch || (await getDefaultBranch()),
  });

  // GitHub returns file content as Base64 — decode it
  return Buffer.from(data.content, "base64").toString("utf-8");
}

// -----------------------------------------------------------
// listFiles(path, branch) — List files in a directory.
//
// Returns an array of { name, path, type } objects.
// type is "file" or "dir".
// -----------------------------------------------------------
export async function listFiles(path = "", branch) {
  const { owner, repo } = getRepoInfo();

  const { data } = await octokit.rest.repos.getContent({
    owner,
    repo,
    path,
    ref: branch || (await getDefaultBranch()),
  });

  return Array.isArray(data)
    ? data.map((item) => ({
        name: item.name,
        path: item.path,
        type: item.type,
      }))
    : [{ name: data.name, path: data.path, type: data.type }];
}

// -----------------------------------------------------------
// createPullRequest({ title, body, head, base })
//
// Opens a new Pull Request on the repo.
//
// PARAMETERS:
//   title — PR title (short, descriptive)
//   body  — PR description (supports Markdown)
//   head  — the branch with your changes
//   base  — the branch you want to merge into (usually "main")
//
// RETURNS: The created PR object (includes .number, .html_url)
// -----------------------------------------------------------
export async function createPullRequest({ title, body, head, base }) {
  const { owner, repo } = getRepoInfo();
  const baseBranch = base || (await getDefaultBranch());

  const { data: pr } = await octokit.rest.pulls.create({
    owner,
    repo,
    title,
    body,
    head,
    base: baseBranch,
  });

  console.log(`✅ PR created: #${pr.number} — ${pr.html_url}`);
  return pr;
}

// -----------------------------------------------------------
// addComment(issueNumber, body) — Post a comment on an issue or PR.
//
// In GitHub's API, issues and PRs share the same comment system.
// So this works for both.
// -----------------------------------------------------------
export async function addComment(issueNumber, body) {
  const { owner, repo } = getRepoInfo();

  await octokit.rest.issues.createComment({
    owner,
    repo,
    issue_number: issueNumber,
    body,
  });

  console.log(`✅ Comment added to #${issueNumber}`);
}

// -----------------------------------------------------------
// addLabels(issueNumber, labels) — Add labels to an issue or PR.
//
// labels is an array of label name strings, e.g. ["bug", "ai-fix"].
// If a label doesn't exist on the repo, GitHub will create it.
// -----------------------------------------------------------
export async function addLabels(issueNumber, labels) {
  const { owner, repo } = getRepoInfo();

  await octokit.rest.issues.addLabels({
    owner,
    repo,
    issue_number: issueNumber,
    labels,
  });

  console.log(`✅ Labels added to #${issueNumber}: ${labels.join(", ")}`);
}

// -----------------------------------------------------------
// getIssue(issueNumber) — Fetch full details of a GitHub issue.
// -----------------------------------------------------------
export async function getIssue(issueNumber) {
  const { owner, repo } = getRepoInfo();

  const { data } = await octokit.rest.issues.get({
    owner,
    repo,
    issue_number: issueNumber,
  });

  return data;
}

// -----------------------------------------------------------
// getPullRequest(prNumber) — Fetch full details of a PR.
// -----------------------------------------------------------
export async function getPullRequest(prNumber) {
  const { owner, repo } = getRepoInfo();

  const { data } = await octokit.rest.pulls.get({
    owner,
    repo,
    pull_number: prNumber,
  });

  return data;
}

// -----------------------------------------------------------
// getPullRequestDiff(prNumber) — Get the raw diff of a PR.
//
// This returns the unified diff text showing exactly what
// lines were added/removed. Used by the PR review agent.
// -----------------------------------------------------------
export async function getPullRequestDiff(prNumber) {
  const { owner, repo } = getRepoInfo();

  const { data } = await octokit.rest.pulls.get({
    owner,
    repo,
    pull_number: prNumber,
    mediaType: { format: "diff" },
  });

  return data;
}

// -----------------------------------------------------------
// getPullRequestFiles(prNumber) — List files changed in a PR.
// -----------------------------------------------------------
export async function getPullRequestFiles(prNumber) {
  const { owner, repo } = getRepoInfo();

  const { data } = await octokit.rest.pulls.listFiles({
    owner,
    repo,
    pull_number: prNumber,
  });

  return data;
}

// -----------------------------------------------------------
// createReviewComment({ prNumber, body, path, line })
//
// Posts an inline review comment on a specific line of a PR.
// This is what shows up as "review comments" in the PR diff view.
// -----------------------------------------------------------
export async function createReviewComment({ prNumber, body, path, line }) {
  const { owner, repo } = getRepoInfo();

  await octokit.rest.pulls.createReviewComment({
    owner,
    repo,
    pull_number: prNumber,
    body,
    path,
    line,
    side: "RIGHT",
    commit_id: (await getPullRequest(prNumber)).head.sha,
  });
}

// -----------------------------------------------------------
// submitReview({ prNumber, body, event })
//
// Submits a full PR review (APPROVE, REQUEST_CHANGES, or COMMENT).
// -----------------------------------------------------------
export async function submitReview({ prNumber, body, event }) {
  const { owner, repo } = getRepoInfo();

  await octokit.rest.pulls.createReview({
    owner,
    repo,
    pull_number: prNumber,
    body,
    event, // "APPROVE" | "REQUEST_CHANGES" | "COMMENT"
  });

  console.log(`✅ Review submitted on PR #${prNumber}: ${event}`);
}

// -----------------------------------------------------------
// mergePullRequest(prNumber, method) — Merge a PR.
//
// method: "merge" | "squash" | "rebase"
// -----------------------------------------------------------
export async function mergePullRequest(prNumber, method = "squash") {
  const { owner, repo } = getRepoInfo();

  await octokit.rest.pulls.merge({
    owner,
    repo,
    pull_number: prNumber,
    merge_method: method,
  });

  console.log(`✅ PR #${prNumber} merged via ${method}`);
}

// -----------------------------------------------------------
// listOpenPullRequests() — List all open PRs in the repo.
// -----------------------------------------------------------
export async function listOpenPullRequests() {
  const { owner, repo } = getRepoInfo();

  const { data } = await octokit.rest.pulls.list({
    owner,
    repo,
    state: "open",
    sort: "updated",
    direction: "desc",
  });

  return data;
}

// -----------------------------------------------------------
// closePullRequest(prNumber) — Close a PR without merging.
// -----------------------------------------------------------
export async function closePullRequest(prNumber) {
  const { owner, repo } = getRepoInfo();

  await octokit.rest.pulls.update({
    owner,
    repo,
    pull_number: prNumber,
    state: "closed",
  });

  console.log(`✅ PR #${prNumber} closed`);
}

// Export the raw octokit instance for advanced use cases
export { octokit };

File: scripts/lib/github_client.py
# ============================================================
# SHARED GITHUB CLIENT — Talks to GitHub API via PyGithub.
#
# WHAT IT PROVIDES:
#   - get_repo()             → PyGithub repo object
#   - get_issue(number)      → issue object
#   - get_file_content(path) → file content as string
#   - list_files(path)       → list of files in a directory
#   - create_branch(name)    → create a new branch
#   - create_or_update_file() → commit a file to a branch
#   - create_pull_request()  → open a PR
#   - add_comment()          → post a comment on issue/PR
#   - add_labels()           → add labels to issue/PR
#   - get_pull_request()     → PR object
#   - get_pr_diff()          → raw diff text
#   - get_pr_files()         → list of changed files
#   - submit_review()        → post a PR review
#   - merge_pr()             → merge a PR
#   - close_pr()             → close a PR
#   - list_open_prs()        → all open PRs
#
# USAGE:
#   from lib.github_client import get_repo, get_issue, create_branch
# ============================================================

import os
import base64
import requests
from github import Github, GithubException
from lib.logger import create_logger

# Create logger for GitHub operations
log = create_logger("GitHub-API")

# -----------------------------------------------------------
# Initialize PyGithub with the token from environment.
# -----------------------------------------------------------
GITHUB_TOKEN = os.environ.get("GITHUB_TOKEN", "")
REPO_NAME = os.environ.get("GITHUB_REPOSITORY", "")

gh = Github(GITHUB_TOKEN)


def get_repo():
    """Get the PyGithub repo object."""
    return gh.get_repo(REPO_NAME)


# -----------------------------------------------------------
# ISSUE OPERATIONS
# -----------------------------------------------------------

def get_issue(issue_number: int):
    """Fetch a GitHub issue by number."""
    repo = get_repo()
    return repo.get_issue(number=issue_number)


# -----------------------------------------------------------
# FILE OPERATIONS — Read files from the repo
# -----------------------------------------------------------

def get_file_content(file_path: str, branch: str = None) -> str:
    """
    Read a file's content from the repo.
    Returns the decoded UTF-8 content.
    """
    repo = get_repo()
    ref = branch or repo.default_branch
    content = repo.get_contents(file_path, ref=ref)
    return base64.b64decode(content.content).decode("utf-8")


def list_files(dir_path: str = "", branch: str = None) -> list:
    """
    List files in a directory.
    Returns list of dicts: [{"name": "app.js", "path": "src/app.js", "type": "file"}]
    """
    repo = get_repo()
    ref = branch or repo.default_branch
    contents = repo.get_contents(dir_path, ref=ref)
    return [
        {"name": item.name, "path": item.path, "type": item.type}
        for item in contents
    ]


# -----------------------------------------------------------
# BRANCH OPERATIONS
# -----------------------------------------------------------

def create_branch(branch_name: str, from_branch: str = None) -> str:
    """
    Create a new branch from the base branch.
    Equivalent to: git checkout -b branch_name
    """
    repo = get_repo()
    base = from_branch or repo.default_branch

    # Get the SHA of the latest commit on the base branch
    base_ref = repo.get_git_ref(f"heads/{base}")
    sha = base_ref.object.sha

    # Create the new branch pointing to that commit
    repo.create_git_ref(ref=f"refs/heads/{branch_name}", sha=sha)
    log.success(f"Branch created: {branch_name} (from {base})")
    return branch_name


# -----------------------------------------------------------
# FILE COMMIT OPERATIONS
# -----------------------------------------------------------

def create_or_update_file(branch: str, file_path: str, content: str, commit_message: str):
    """
    Commit a file to a branch via GitHub API.
    Creates the file if it doesn't exist, updates if it does.
    """
    repo = get_repo()

    # Check if file already exists (need its SHA to update)
    try:
        existing = repo.get_contents(file_path, ref=branch)
        repo.update_file(
            path=file_path,
            message=commit_message,
            content=content,
            sha=existing.sha,
            branch=branch,
        )
    except GithubException as e:
        if e.status == 404:
            # File doesn't exist — create it
            repo.create_file(
                path=file_path,
                message=commit_message,
                content=content,
                branch=branch,
            )
        else:
            raise

    log.success(f"File committed: {file_path} → {branch}")


# -----------------------------------------------------------
# PULL REQUEST OPERATIONS
# -----------------------------------------------------------

def create_pull_request(title: str, body: str, head: str, base: str = None) -> dict:
    """
    Open a new Pull Request.
    Returns dict with: number, html_url, title
    """
    repo = get_repo()
    base_branch = base or repo.default_branch

    pr = repo.create_pull(title=title, body=body, head=head, base=base_branch)
    log.success(f"PR created: #{pr.number} — {pr.html_url}")

    return {"number": pr.number, "html_url": pr.html_url, "title": pr.title}


def get_pull_request(pr_number: int):
    """Fetch a PR object by number."""
    repo = get_repo()
    return repo.get_pull(pr_number)


def get_pr_diff(pr_number: int) -> str:
    """
    Get the raw unified diff of a PR.
    Uses the GitHub REST API directly since PyGithub doesn't support diff format.
    """
    url = f"https://api.github.com/repos/{REPO_NAME}/pulls/{pr_number}"
    headers = {
        "Authorization": f"token {GITHUB_TOKEN}",
        "Accept": "application/vnd.github.v3.diff",
    }
    response = requests.get(url, headers=headers)
    response.raise_for_status()
    return response.text


def get_pr_files(pr_number: int) -> list:
    """
    List files changed in a PR.
    Returns list of dicts with: filename, additions, deletions, status
    """
    repo = get_repo()
    pr = repo.get_pull(pr_number)
    return [
        {
            "filename": f.filename,
            "additions": f.additions,
            "deletions": f.deletions,
            "status": f.status,
        }
        for f in pr.get_files()
    ]


def submit_review(pr_number: int, body: str, event: str):
    """
    Submit a PR review.
    event: "APPROVE", "REQUEST_CHANGES", or "COMMENT"
    """
    repo = get_repo()
    pr = repo.get_pull(pr_number)
    pr.create_review(body=body, event=event)
    log.success(f"Review submitted on PR #{pr_number}: {event}")


def merge_pr(pr_number: int, method: str = "squash"):
    """
    Merge a PR.
    method: "merge", "squash", or "rebase"
    """
    repo = get_repo()
    pr = repo.get_pull(pr_number)
    pr.merge(merge_method=method)
    log.success(f"PR #{pr_number} merged via {method}")


def close_pr(pr_number: int):
    """Close a PR without merging."""
    repo = get_repo()
    pr = repo.get_pull(pr_number)
    pr.edit(state="closed")
    log.success(f"PR #{pr_number} closed")


def list_open_prs() -> list:
    """List all open PRs in the repo."""
    repo = get_repo()
    return list(repo.get_pulls(state="open", sort="updated", direction="desc"))


# -----------------------------------------------------------
# COMMENT & LABEL OPERATIONS
# -----------------------------------------------------------

def add_comment(issue_number: int, body: str):
    """Post a comment on an issue or PR."""
    repo = get_repo()
    issue = repo.get_issue(number=issue_number)
    issue.create_comment(body=body)
    log.success(f"Comment added to #{issue_number}")


def add_labels(issue_number: int, labels: list):
    """Add labels to an issue or PR."""
    repo = get_repo()
    issue = repo.get_issue(number=issue_number)

    for label_name in labels:
        try:
            repo.get_label(label_name)
        except GithubException:
            # Label doesn't exist — create it
            repo.create_label(name=label_name, color="7057ff")

    issue.add_to_labels(*labels)
    log.success(f"Labels added to #{issue_number}: {', '.join(labels)}")
